File: server/foxglove/messages/twist_stamped.py
from dataclasses import dataclass
import json
import struct


@dataclass
class TwistStamped:
    seconds: int
    nanoseconds: int
    linear_x: float
    linear_y: float
    linear_z: float
    angular_x: float
    angular_y: float
    angular_z: float

    def to_frame(self, channel_id: int, frame_id: str):
        data = {
            "header": {
                "stamp": {"sec": self.seconds, "nanosec": self.nanoseconds},
                "frame_id": frame_id,
            },
            "twist": {
                "linear": {"x": self.linear_x, "y": self.linear_y, "z": self.linear_z},
                "angular": {
                    "x": self.angular_x,
                    "y": self.angular_y,
                    "z": self.angular_z,
                },
            },
        }
        payload = json.dumps(data).encode("utf-8")
        frame = b"\x01" + struct.pack("<I", channel_id) + payload
        return frame

File: web/src/foxglove/types/twist-stamped.ts
export interface TwistStamped {
	linearX: number;
	linearY: number;
	linearZ: number;
	angularX: number;
	angularY: number;
	angularZ: number;
	seconds: number;
	nanoseconds: number;
}

export function twistStampedFrame(
	twistStamped: TwistStamped,
	channelId: number,
	frameId: string,
) {
	const data = {
		header: {
			stamp: {
				sec: twistStamped.seconds,
				nanosec: twistStamped.nanoseconds,
			},
			frame_id: frameId,
		},
		twist: {
			linear: {
				x: twistStamped.linearX,
				y: twistStamped.linearY,
				z: twistStamped.linearZ,
			},
			angular: {
				x: twistStamped.angularX,
				y: twistStamped.angularY,
				z: twistStamped.angularZ,
			},
		},
	};

	const dataStr = JSON.stringify(data);
	const encoder = new TextEncoder();
	const dataBytes = encoder.encode(dataStr);
	const buffer = new ArrayBuffer(1 + 4 * dataBytes.length);
	const view = new DataView(buffer);
	view.setUint8(0, 0x01);
	view.setUint32(1, channelId, true);
	const result = new Uint8Array(buffer);
	result.set(dataBytes, 5);
	return result;
}
